File: Second Year/PDOO/Napakalaki/src/NapakalakiGame/Player.java
/*  
 *  Programación y diseño orientado a objetos
 *  Grado en Ingeniería Informática
 * 
 *  
 *
 *  Angela Castro Jiménez
 *  Francisco Carrillo Pérez
 */
package NapakalakiGame;
import java.util.ArrayList;

public class Player {
   private boolean dead = true;
   private String name;
   private int level;
   
   private ArrayList<Treasure> visibleTreasures = new ArrayList();
   private ArrayList<Treasure> hiddenTreasures = new ArrayList();
   private BadConsequence pendingBadConsequence = new BadConsequence();
   
   public Player(String miname){
       name = miname;
       level = 1;
   }

   public Player(Player p)
   {
       this.name = p.name;
       this.dead = p.dead;
       this.level = p.level;
       
       
   }
   public void bringTolife(){
       dead = false;        
   }
   public boolean getDeath()
   {
       return dead;
   }
   //Devuelve el nivel de combate del jugador, que viene dado por su nivel más 
    //los bonus que le proporcionan los tesoros que tenga equipados, según las 
    //reglas del juego.
     public int getCombatLevel(){
         boolean tiene = false;
         int levelfinal=level;
         for(Treasure t: visibleTreasures){
             if(t.getType()==TreasureKind.NECKLACE){
                tiene = true;
             }
         }
         
         for(Treasure t: visibleTreasures){
             if(tiene){
                 levelfinal=levelfinal+ t.getMaxBonus();
             }
             else{
                 levelfinal=levelfinal+ t.getMinBonus();
             }
         
         }
         return levelfinal;      
     
     }        
   
   public void incrementLevels(int I){
      level = level + I;
   }
   
   public void decrementLevels(int I){
       if(level<=I)
           level=1;
       else
        level = level - I;
   }
   
   public void setPendingBadConsequence(BadConsequence b){
      pendingBadConsequence = b;
   }
   
   //Cambia el estado de jugador a muerto si no tiene ni tesoros visibles ni 
    //ocultos, modificandon el correspondiente atributo.
   public void dieIfNoTreasures() {
      if(visibleTreasures.size() <= 0 && hiddenTreasures.size() <= 0){
          dead = true;
      }
   }
   
   //Si el jugador tiene equipado el tesoro tipo NECKLACE, se lo pasa al CardDealer y lo
   //elimina de sus tesoros visibles.
   public void discardNecklaceVisible() {
         for (Treasure t : this.visibleTreasures) {
 
            if (t.getType() == TreasureKind.NECKLACE) {
 
                //Le pasamos el tesoro al CardDealer
                CardDealer cd = CardDealer.getInstance();
                cd.giveTreasuresBack(t);
 
                //Lo eliminamos de nuestros tesoros visibles (equipados)
                visibleTreasures.remove(t);
 
                //Salimos del bucle
                break;
            }
         }
   }
  
   
   public void die() {
       this.setLevels(1);
       CardDealer dealer = CardDealer.getInstance();
       for(Treasure t:visibleTreasures)
       {
           dealer.giveTreasuresBack(t);
       }
       visibleTreasures.clear();
       for(Treasure t: hiddenTreasures)
       {
           dealer.giveTreasuresBack(t);
       }
       hiddenTreasures.clear();
       this.dieIfNoTreasures();
       
   }
   
   // Calcula y devuelve los niveles que puede comprar el jugador con la lista t de tesoros.
   protected int computeGoldCoinsValue( ArrayList<Treasure> t) {
        int suma = 0;
        for(Treasure tr: t){
            suma = suma + tr.getGoldCoins();
        }
        return suma%1000;
   }
   
 
   public boolean canBuyLevels(int l){
       return this.level+l<10;

   }                         
   
   
   public void applyPrize(Monster currentMonster){
       int nLevels =  currentMonster.getLevelsGained();
       this.incrementLevels(nLevels);
       int nTreasures = currentMonster.getTreasuresGained();
       if(nTreasures >0)
       {
           CardDealer dealer = CardDealer.getInstance();
           int i;
           for(i = 1; i <= nTreasures;i++)
           {
               Treasure treasure = dealer.nextTreasure();
               hiddenTreasures.add(treasure);
           }
       }
   }
   
   
   public void applyBadConsequence(BadConsequence bad){
       int nLevels = bad.getLevels();
       this.decrementLevels(nLevels);
       BadConsequence pendingBad = bad.adjustToFitTreasureLists(visibleTreasures, hiddenTreasures);
       setPendingBadConsequence(pendingBad);
   }
   public BadConsequence getPendingBadConsequence()
   {
       return pendingBadConsequence;
   }
    private boolean isTreasureKindInUse(TreasureKind type) {
        boolean result = false;
        for (Treasure tv : this.visibleTreasures) {
 
            if (type.equals(tv.getType())) {
 
                result = true;
                break;
 
            }
 
        }
        return result;
    }
   //Comprueba si el tesoro (t) se puede pasar de oculto a visible, 
    //según las reglas del juego
   public boolean canMakeTreasureVisible(Treasure t){
       if(t.getType()==TreasureKind.ONEHAND){
           int cont=0;
           for(Treasure tr: visibleTreasures){
               if(tr.getType()==TreasureKind.ONEHAND)
                   cont++;
               if(tr.getType()==TreasureKind.BOTHHAND)
                   return false;
           }
       
       }
       else if (t.getType()==TreasureKind.BOTHHAND){
           for(Treasure tr: visibleTreasures){
               if(tr.getType()==TreasureKind.ONEHAND || tr.getType()==TreasureKind.BOTHHAND)
                   return false;
           }
       }
        else{
           for(Treasure tr: visibleTreasures){
               if(tr.getType()==t.getType())
                   return false;
           }
       }
       return true;
       
    }

   
   public int howManyVisibleTreasures(TreasureKind tkind){
       int contador = 0;
       for(int i = 0; i < visibleTreasures.size();i++){
          Treasure t;
          t = visibleTreasures.get(i);
          if(t.getType() == tkind){
              contador++;
          }
        }
       return contador;     
    }
   
   public boolean isDead(){
       return dead;
   }
   
   
  public ArrayList<Treasure> getHiddenTreasures(){
     
       
       return hiddenTreasures;
  }
   
   
   public ArrayList<Treasure> getVisibleTreasures(){
       
       return visibleTreasures;
   }
   
   
  public CombatResult combat(Monster m){
      CombatResult combatResult;
      
      int myLevels = this.getCombatLevel();
      int monsterlevel = this.getOponentLevel(m);
      if(myLevels > monsterlevel){
          this.applyPrize(m);
          if(this.level >=10){
              combatResult = CombatResult.WinAndWinGame;
          }
          else{
              combatResult = CombatResult.Win;
          }
      }
       else{
          Dice dice = Dice.getInstance();
          int escape = dice.nextNumber();
          if(escape<5){
              boolean amIDead = m.kills();
              if(amIDead == true) {
                  this.die();
                  combatResult=CombatResult.LoseAndDie;
              }
              else {
                  boolean should = this.shouldConvert();
                  if(should)
                  {
                      combatResult=CombatResult.LoseAndConvert;
                  }
                  else
                  {
                       BadConsequence bad = m.getBadConsequence();
                       combatResult=CombatResult.Lose;
                       this.applyBadConsequence(bad);
                  }
                 
              }
          }
          else{
              combatResult=CombatResult.LoseAndEscape;
          }
        }
      
      this.discardNecklaceVisible();
      return combatResult;
  }
   
  
   public void makeTreasureVisible(Treasure t) {
       boolean canl = this.canMakeTreasureVisible(t);
       if(canl)
       {
           visibleTreasures.add(t);
           hiddenTreasures.remove(t);
       }
   }
  
   public void discardVisibleTreasure(Treasure t) {
       visibleTreasures.remove(t);
       if((pendingBadConsequence != null) && (!pendingBadConsequence.isEmpty())){
           pendingBadConsequence.substractVisibleTreasure(t);
       }
       this.dieIfNoTreasures();
       
   }
   
   
  public void discardHiddenTreasure(Treasure t){
      hiddenTreasures.remove(t);
      if((pendingBadConsequence != null) && (!pendingBadConsequence.isEmpty())){
          pendingBadConsequence.substractHiddenTreasure(t);
      }
      this.dieIfNoTreasures();
  }
   
   
   public boolean buyLevels(ArrayList<Treasure> visible,ArrayList<Treasure> hidden){
        float levelsMayBought=computeGoldCoinsValue(visible);
        levelsMayBought+=computeGoldCoinsValue(hidden);
        int levels=(int)levelsMayBought/1000;
        boolean canI=canBuyLevels(levels);
        if(canI)
            incrementLevels(levels);
        visibleTreasures.removeAll(visible);
        hiddenTreasures.removeAll(hidden);
        CardDealer dealer=CardDealer.getInstance();
        for(Treasure t:visible)
            dealer.giveTreasuresBack(t);
        for(Treasure t:hidden)
            dealer.giveTreasuresBack(t);
        return canI;
       
   }
   
   public boolean validState(){
     return (pendingBadConsequence.isEmpty() &&  hiddenTreasures.size() <= 4);
   }
   
   
   public void initTreasures() {
       CardDealer dealer = CardDealer.getInstance();
       Dice dice = Dice.getInstance();
       this.bringTolife();
       Treasure treasure = dealer.nextTreasure();
       hiddenTreasures.add(treasure);
       int number = dice.nextNumber();
       if(number  > 1)
       {
           treasure = dealer.nextTreasure();
           hiddenTreasures.add(treasure);
       }
       if(number == 6)
       {
           treasure = dealer.nextTreasure();
           hiddenTreasures.add(treasure);
       }
   }
   
   public boolean hasVisibleTreasures(){
     return(visibleTreasures.size() > 0);
   }
   
   public int getLevels(){
       return level;
   }
   
   public void setLevels(int i)
   {
       level = i;
   }
   
   public String getName()
   {
       return name;
   }
 
   
   protected boolean shouldConvert()
    {
        Dice dice = Dice.getInstance();
        int escape = dice.nextNumber();
        
        return (escape == 6);
       
    }
    
   protected int getOponentLevel(Monster m)
    {
        return m.getCombatLevel();
    }
   
   //EXAMEN
     public void changeTreasure(Treasure t)
    {
        CardDealer dealer = CardDealer.getInstance();
        Treasure treasure = dealer.nextTreasure();
        hiddenTreasures.remove(t);
        hiddenTreasures.add(treasure);
        
        
    }
     //FIN EXAMEN
  public String toString(){
        return "Nombre: "+name
                +" \nNivel del jugador " + Integer.toString(level)
                +"\nNivel de combate: "+ Integer.toString(getCombatLevel());
     }
}

File: Second Year/PDOO/Napakalakirb/lib/player.rb
# encoding: UTF-8
#
#   Programación y diseño orientado a objetos
#   Grado en Ingeniería Informática
#
#  Ángela Castro Jimenez
#  Francisco Carrillo Pérez
module Napakalaki
  require_relative 'BadConsequence.rb'
  require_relative 'Prize.rb'
  require_relative 'dice'
  
  require_relative 'card_dealer'
  require_relative 'TreasureKind'
  
#  require 'singleton'

class Player
#    include Singleton 
    attr_accessor :name, :dead,:level, :pendingBadConsequence,:visibleTreasures,:hiddenTreasures
    
     
      
#  private_class_method :new
   
    
   def initialize (name)
     if(name.class.eql?(String))
      @name = name
      @dead = true
      @level = 1
      @hiddenTreasures = Array.new
      @visibleTreasures = Array.new
      @pendingBadConsequence = BadConsequence.newLevelNumberOfTreasures("", 0, 0, 0)
     
     else
     @name = name.name
     @dead = name.dead
     @level = name.level
     @pendingBadConsequence = name.pendingBadConsequence
     @visibleTreasures = name.visibleTreasures
     @hiddenTreasures = name.hiddenTreasures
     end
   end
   
   def oponent_Level(m)
      return m.getBasicValue
  end
  
   def should_convert
    dice=Dice.instance
    return dice.nextNumber==6
  end
  
   def to_s
      "#{@name}\n Level: #{@level}\n"
  end
   def bringTolife()
     @dead = false
   end
   
   def incrementLevels(i)
       @level = @level + i
   end
   
   def decrementLevels (i)
     @level = @level - i
   end
   
   def dieIfNoTreasures()
       if (@visibleTreasures.size <= 0 && @hiddenTreasures.size <= 0)
            @dead = true
       end
   end
#  //Devuelve el nivel de combate del jugador, que viene dado por su nivel más 
#    //los bonus que le proporcionan los tesoros que tenga equipados, según las 
#    //reglas del juego.
   def getCombatLevel()
      tiene = false
      levelfinal = @level
      @visibleTreasures.each{|t|
        if t.type == TreasureKind::NECKLACE
          tiene = true
        end
      }
       @visibleTreasures.each{|t|
         if tiene
           levelfinal = levelfinal + t.getMaxBonus()
         else
           levelfinal = levelfinal + t.getMinBonus()
         end
       }
       
      return levelfinal
   end
   def incrementLevels(aumento)
      @level = @level + aumento
   end
   
   def decrementLevels(decremento)
     @level = @level - decremento
   end
#Si el jugador tiene equipado el tesoro tipo NECKLACE, se lo pasa al CardDealer y lo
#elimina de sus tesoros visibles. 
   def discardNecklaceIfVisible()
       collar = false;
      for i in 0..@visibleTreasures.size()-1
        if collar then
          break
        end
        if (@visibleTreasures.at(i).type().to_s == "[:NECKLACE]")
          collar = true
          tesoro = @visibleTreasures.at(i)
          CardDealer.instance.giveTreasureBack(tesoro)
          @visibleTreasures.delete(tesoro)
          @pendingBadConsequence.substractVisibleTreasure(tesoro) 
          dielfNoTreasures
        end
      end
        end
 
      
      
   


   def die()
     @level = 1
     dealer = CardDealer.instance
     @visibleTreasures.each{|t|
       dealer.giveTreasureBack(t);
       
     }
     
     @hiddenTreasures.each{|t|
       dealer.giveTreasureBack(t);
       
     }
     @visibleTreasures.clear
     @hiddenTreasures.clear
     dieIfNoTreasures
     
     
   end


   def computeGoldCoinsValue(t)
     suma = 0
     t.each{ |tr|
        suma = suma + tr.getGoldCoins
     }
     
     return suma
     
   end

   def  canBuyLevels(i)
      puedecomprar = false
      sumamonedas = 0
      
     @visibleTreasures.each{ |t|
       sumamonedas = sumamonedas + t.getGoldCoins
       if(sumamonedas >= i*1000)
         puedecomprar = true
       end
     }
     if (puedecomprar)
        return puedecomprar
     
     else
       @hiddenTreasures.each{ |t|
         sumamonedas = sumamonedas + t.getGoldCoins
       if(sumamonedas >= i*1000)
         puedecomprar = true
       end
       }
       return puedecomprar
   
    end
   
  end
  
 
   def applyPrize(currentMonster)

    nLevels = currentMonster.getLevelsGained()
    incrementLevels(nLevels)
    nTreasures = currentMonster.getTreasuresGained
        
      if(nTreasures > 0)
        dealer = CardDealer.instance
        for i in 0..(nTreasures-1)
          t = dealer.nextTreasure
          @hiddenTreasures << t
        end
    end
   end


   def applyBadConsequence(bad)
     nLevels = bad.getLevels
     decrementLevels(nLevels)
     pendingBad = bad.adjustToFitTreasureLists(@visibleTreasures, @hiddenTreasures)
     setPendingBadConsequence(pendingBad)
   end


   
   def canMakeTreasureVisible(treasure)
     isWearable=false;
      tipo=treasure.getType.to_s
      contador=howManyVisibleTreasures(tipo)
      
      if(tipo.eql?("[:ONEHAND]") && contador<2 && howManyVisibleTreasures("[:BOTHHAND]")==0)
        isWearable=true
      elsif(tipo.eql?("[:BOTHHAND]") && contador==0 && howManyVisibleTreasures("[:ONEHAND]")==0)
        isWearable=true
      elsif(tipo!="[:BOTHHAND]" && tipo!="[:ONEHAND]" && contador==0)
        isWearable=true
      end
      
      return isWearable 
   
    end
    
   


   def howManyVisibleTreasures(tKind)
     contador = 0
     @visibleTreasures.each{ |t|
       if(t.getType == tKind)
         contador = contador + 1
       end
     }
     return contador
   end
   
   def isDead()
      @dead
   end
   
#   Aun no implementado
   def getHiddenTreasures()
     @hiddenTreasures
   end

#   Aun no implementado
   def getVisibleTreasures()
     @visibleTreasures
   end
   
def getName
  return @name
end


    def combat(m)
    myLevel = getCombatLevel
    monsterLevel = m.getCombatLevel
        
    if(myLevel > monsterLevel)
      applyPrize(m);
      if(Napakalaki.instance.currentPlayer.level >= 10)
          combatResult = CombatResult::WINANDWINGAME
      else
          combatResult = CombatResult::WIN
      end
    else
      escape = Dice.instance().nextNumber()
        if(escape < 5)
          amIDead = m.kills
          if(amIDead)
              die
              combatResult = CombatResult::LOSEANDDIE
          elsif should_convert then
                    
            combatResult = CombatReesult::LOSEANDCONVERT
          else
            combatResult = CombatResult::LOSE
            bad = m.getBadConsequence
            applyBadConsequence(bad)
          end
        else
          combatResult = CombatResult::LOSEANDESCAPE
        end
    end
    discardNecklaceIfVisible
     return combatResult
  end

   def isTreasureKindInUse(type) 
 
      result = false
      @visibleTreasures.each { |tv|
        if type == tv.type then
 
          result = true
          break
          end
 
      }
 
      
      return result
         
    end

   def makeTreasureVisible(t)
     canl = canMakeTreasureVisible(t)
     if canl
       @visibleTreasures << t
       @hiddenTreasures.delete(t)
     end
    
   end
  
   def discardVisibleTreasure(t)
      @visibleTreasures.delete(t)
      @pendingBadConsequence.substractVisibleTreasure(t) 
      dielfNoTreasures
   end


   def discardHiddenTreasure(t)
     @hiddenTreasures.delete(t)
      @pendingBadConsequence.substractHiddenTreasure(t)
      dieIfNoTreasures 
   end


   def buyLevels(visible,hidden)
     levelsMayBought = computeGoldCoinsValue(visible)
     levelsMayBought = levelsMayBought + computeGoldCoinsValue(hidden)
     levels = levelsMayBought / 1000
     canl = canBuyLevels(levels)
     if canl
       incrementLevels(levels)
     end
    
     dealer = CardDealer.instance
     visible.each{|t|
       dealer.giveTreasureBack(t)
       @visibleTreasures.delete(t)
       
     }
     hidden.each{|t|
       dealer.giveTreasureBack(t)
       @hiddenTreasures.delete(t)
       
     }
     return canl;
   end

   
  def validState
    if(@pendingBadConsequence.isEmpty && @hiddenTreasures.size <= 4)
      return true
    else
      return false
    end
  end


  def initTreasures()
    dealer = CardDealer.instance
    dice = Dice.instance
    bringTolife
    treasure = dealer.nextTreasure
    @hiddenTreasures << treasure
    number = dice.nextNumber
    if number > 1
       treasure = dealer.nextTreasure
       @hiddenTreasures << treasure
    end
    if number == 6
      treasure = dealer.nextTreasure
      @hiddenTreasures << treasure
      
    end
  end

  def hasVisibleTreasures()
    if(@visibleTreasures.size > 0)
      return true
    else
      return false
    end
  end
  
   #EXAMEN
   def addHiddenTreasureExamen(t)
      @hiddenTreasures << t
   end
   #FIN EXAMEN
end
 


end

